### ISSUE TITLE
"IAM Authentication token has expired" while using driver with IAM authentication

### ISSUE THREAD

**Issue by vitgorbunov** [aws/amazon-redshift-jdbc-driver#97]
## Driver version
2.1.0.9

## Redshift version
1.0.54052

## Client Operating System
Java 11 lambda runtime

## JAVA/JVM version
Java 11 lambda runtime

## Problem description
Sometimes when our app is trying to create new connection it fails with "FATAL: IAM Authentication token has expired"
1. Expected behaviour: Able to create jdbc connection
2. Actual behaviour: JDBC connection sometimes fails with IAM Authentication token has expired.
3. Error message/stack trace:
```
Caused by: java.sql.SQLException: FATAL: IAM Authentication token has expired
at com.amazon.redshift.util.RedshiftException.getSQLException(RedshiftException.java:56)
at com.amazon.redshift.Driver.connect(Driver.java:339)
at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
at com.tealium.redshift.automation.services.redshift.JDBCConnectionFactory.createConnection(JDBCConnectionFactory.java:20)
at com.tealium.redshift.automation.services.redshift.JDBCConnectionFactory_ClientProxy.createConnection(Unknown Source)

```
4. Any other details that can be helpful:

In class 'IamHelper' it caches credentials in HashMap:private static Map<String, GetClusterCredentialsResult> credentialsCache = new HashMap<String, GetClusterCredentialsResult>();and in the row 636 and similar places it has the following check:
```java
if (credentials == null || (providerType == CredentialProviderType.PLUGIN && idpCredentialsRefresh)
  || credentials.getExpiration().before(new Date(System.currentTimeMillis() - 60 * 1000 * 5))) {
```
So at client side it considers that credentials from cache are usable additional 5 mins from expiration time.

## Reproduction code

Below is a code snippet that uses the logic from jdbc driver and demonstrates that expired token is not considered expired using this code:

```java
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

class Scratch {
    public static void main(String[] args) throws ParseException {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");

        Date currentDate = dateFormat.parse("2023-09-19 12:00:00.000");
        Date expirationDate = dateFormat.parse("2023-09-19 11:56:00.000");

        System.out.println(expirationDate.before(new Date(currentDate.getTime() - 60 * 1000 * 5))); //gives false
        //potential fix
        System.out.println(expirationDate.before(new Date(currentDate.getTime() + 60 * 1000 * 5))); // gives true
    }
}
```

**Comment by bhvkshah** [comment]
Hi @vitgorbunov , thanks for opening this issue and bringing this to our attention! I will take a look and get abck to you when I have an update.

**Comment by bhvkshah** [comment]
@vitgorbunov I don't see anything wrong with current logic - we check if the token expired more than 5 mins ago, and if so you get an expired token error.

Were you intending to convey something else?

**Comment by vitgorbunov** [comment]
@bhvkshah If token has expired 1-4 minutes ago - cache will still consider it valid, and as a result the connection to be created will have invalid token.

We used to simply make the following call to get connection.
DriverManager.getConnection(String.format("jdbc:redshift:iam://%s:%s/%s", clusterId, region, db), user, null);

As you see I don't specify any password. It's generated by driver under the hood and cached, thus it's very confusing to get "IAM Authentication token has expired".  And I can't really retry the call until 5 minutes since token expired, because expired token is not removed from cache.

Now we have to maintain our own cache of temporary credentials and pass them to getConnection method.

**Comment by bhvkshah** [comment]
I see your concern. We would have to test this fix for the use case of expiring credentials. Would you prefer to raise a PR for this yourself?

**Comment by bhvkshah** [comment]
@vitgorbunov The fix for this was released in driver version 2.1.0.21. Thank you for your contribution to the Redshift JDBC Driver!

**Comment by vitgorbunov** [comment]
Thank you @bhvkshah !
